### ISSUE TITLE
Uncaught (in promise) FirebaseError: Missing or insufficient permissions.

### ISSUE THREAD

**Issue by ThigoPSI** [actions-on-google/smart-home-nodejs#369]
Hi, can anyone help me fix this error, I'm having a lot of trouble with firebase
![Capturar](https://user-images.githubusercontent.com/50698561/57842029-a0b42780-77c3-11e9-80ff-beac1432a037.JPG)

**Comment by Fleker** [comment]
When you look at your Firestore rules, they should match:

```
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read;
      allow write: if false;
    }
  }
}
```

If you do not allow read, you may see this issue in the client.

**Comment by ThigoPSI** [comment]
this I have the same, I have been more than 1 week and I can not connect to firebase
![Cadpturar](https://user-images.githubusercontent.com/50698561/57860262-7aef4880-77ec-11e9-848b-3d431f1c6320.JPG)

**Comment by Fleker** [comment]
That seems to be a separate issue, unrelated to Firebase permissions, and may be related to an error on the server-side implementation that is causing the server to fail early so that it is not adding the CORS header to the response.

**Comment by ThigoPSI** [comment]
and you think you can help me solve this ??

**Comment by Fleker** [comment]
Are you noticing any errors in the server?

**Comment by ThigoPSI** [comment]
which I do not see, it runs all right, but when I try to add a device of those mistakes I showed up, but be as it is there prints for you to see
![server](https://user-images.githubusercontent.com/50698561/57869701-8139f080-77fd-11e9-9ee6-5f8fbc24d7dd.JPG)

**Comment by Fleker** [comment]
What is 'smart-home-nodejs/frontend/functions'? This project is designed to use App Engine, not cloud functions.

**Comment by Fleker** [comment]
Looking at the left-hand of the screen, the issue may be not including the smart-home-key.json while also running the service through ngrok (#366).

**Comment by guilhermelirio** [comment]
> 
> 
> this I have the same, I have been more than 1 week and I can not connect to firebase
> ![Cadpturar](https://user-images.githubusercontent.com/50698561/57860262-7aef4880-77ec-11e9-848b-3d431f1c6320.JPG)

Same error here! =/

**Comment by mjoelmendes** [comment]
1+ #370  where is the solution?

**Comment by Fleker** [comment]
It seems like there is an issue in your server-side App Engine implementation that is causing the implementation to crash.

**Comment by mjoelmendes** [comment]
and this problem is on our side? or is it something that is wrong in the example?

**Comment by Fleker** [comment]
Without seeing what logs are being sent in your server, it is hard to identify why it is crashing.

**Comment by mjoelmendes** [comment]
Where can I go to get this log?

**Comment by Fleker** [comment]
If it's hosted on App Engine, you can go to console.cloud.google.com/logs/viewer

**Comment by ThigoPSI** [comment]
I use Firebase and the sample servers. How can I solve the problem? and where can I get the log?

**Comment by Fleker** [comment]
If you go to the link above you should be able to view server logs.

**Comment by ThigoPSI** [comment]
is there, do you think you can help me?

[log.zip](https://github.com/actions-on-google/smart-home-nodejs/files/3191771/log.zip)

**Comment by Fleker** [comment]
None of the included logs contain an error

**Comment by ThigoPSI** [comment]
So why on my smart home node server js (NGROk) do not seem like devices that I create on the smart frontend server at home?

**Comment by Fleker** [comment]
You're using ngrok for the server?

**Comment by ThigoPSI** [comment]
yes it is my local server. 

and I want to simulate (add) devices on it

**Comment by Fleker** [comment]
Okay. In that case, the problem is happening because the API calls are being made to a URL that does not exist. You'll need to follow a few steps as described in #366 to update your database connections, and update the API endpoint to your ngrok URL.

**Comment by ThigoPSI** [comment]
is always giving error

[2019-05-17T14_49_45_815Z-debug.zip](https://github.com/actions-on-google/smart-home-nodejs/files/3191979/2019-05-17T14_49_45_815Z-debug.zip)

**Comment by Fleker** [comment]
It seems the error is related to the command `tsc && copy src/*.json dist/`, which does not seem to be running as expected in your environment.

**Comment by radianceltd** [comment]
Error: 7 PERMISSION_DENIED: Missing or insufficient permissions.
    at Object.callErrorFromStatus (/srv/node_modules/firebase-admin/node_modules/@grpc/grpc-js/build/src/call.js:30:26)
    at Http2CallStream.call.on (/srv/node_modules/firebase-admin/node_modules/@grpc/grpc-js/build/src/client.js:96:33)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/firebase-admin/node_modules/@grpc/grpc-js/build/src/call-stream.js:100:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)

**Comment by radianceltd** [comment]
Can you help me out

**Comment by spieker** [comment]
I don't know if this is the issue here, but when creating a new database in firebase, the default permissions are set to "allow everything for 1 week after creation". This looks something like this:

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.time < timestamp.date(2020, 7, 1);
    }
  }
}
```

You can extend the permitted time by modifying `timestamp.date(2020, 7, 1)`, or create propper permission rules.

**Comment by alirehman726** [comment]
service cloud.firestore {
  match /databases/{database}/documents {
    match /Users/{userId} {
    	allow read, update, delete: if request.auth.uid == userId;
      allow create: if request.auth.uid != null;
    }
    match /{document=**} {
    	allow read, write: if request.auth != null;
    }
  }
}

**Comment by alirehman726** [comment]
![Annotation 2020-08-08 153741](https://user-images.githubusercontent.com/67003105/89707746-a381d600-d98e-11ea-882f-e258a2205a5c.png)

**Comment by Taaseen71** [comment]
I had the same error, And after reading everything in the comments section, all i did was change it from 

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.time < timestamp.date(2020, 7, 1);
    }
  }
}
```

to this

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write;
    }
  }
}
```

so basically allow read and write indefinitely

**Comment by Noly17** [comment]
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {

    match /{document=**} {

//this code underneath is the thing. go change the date, i change mine 2020 to 2021
      allow read, write: if request.time < timestamp.date(2021, 8, 29);
    }
  }
}

**Comment by astroana** [comment]
This is only applicable for development but not advisable for production, I also have the same issue whenever I change the rules with auth even though I detached the listener upon logout.

**Comment by astroana** [comment]
> rules_version = '2';
> service cloud.firestore {
>   match /databases/{database}/documents {
>     match /{document=**} {
>       allow read, write;
>     }
>   }
> }


This is only applicable for development and not advisable for production. I have the same issue when changing rules with auth even though I detached the listener upon logout.

**Comment by Rian1010** [comment]
@Taaseen71 Hi! :) Where should it be added please? So, which file please?

**Comment by Oshanwick** [comment]
@alirehman726  its works thanks man

**Comment by kadebedis** [comment]
If this should be opened in a different issue, please let me know. I get the same error that people above are getting, but my application is able to read & write the data to the database. I just want to get the error to go away.

Has anyone else seen this issue?

![image](https://user-images.githubusercontent.com/18197391/97917202-5f923900-1d11-11eb-9e71-25d14717bc32.png)

**Comment by yoonghan** [comment]
@Rian1010 login to https://console.firebase.google.com/, access Cloud Firestore.
From the tabs appeared - Data/Rules/Indexes/Usage. Select "Rules", and change to "allow read, write;". Save it.
Though i believe this will be a security issue in production, but for now it solves my problem.

Thanks @alirehman726  it worked.

**Comment by femzykel** [comment]
I had the same error, But all i did was changing the rule that works with the date and it worked perfectly. The current request.time was  "<" the timestamp.date which invalidates the rule once the date/time elapse the date of the default  rule.

rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.time < timestamp.date(2020, 9, 10);
    }
  }
}
Updated to this

rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.time > timestamp.date(2020, 9, 10);
    }
  }
}
so basically i reversed the lesser-than sign to greater-than sign giving access to my database to function as it should,
or you can just manually keep changing the date to extend permission without switching the the lesser-than sign to greater-than sign.

**Comment by emregurbulak** [comment]
Hi,
I got a same issue .  The problem was about firebase test project time to expired .
Because the test project has a 30 days access permission. 
I was create a new project on firebase and  I took the configuration(firebaseConfig) and replaced it with the old firebaseConfig in my project code.

**Comment by crevulus** [comment]
Hi all, I've changed my rules to be more liberal but I still have the same error when I try to `get()` a collection. My rules are as follows:

```
rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /users/{userId} {
      allow read;
      allow write: if request.auth.uid == userId;
      allow update: if request.auth.uid == userId;
    }
  }
}
```

Previously, `read` followed the same structure as the rules for `write` and `update`. But I still get `Uncaught (in promise) FirebaseError: Missing or insufficient permissions.` even though I'm allowing all reads.

**Comment by imadelcass** [comment]
> I don't know if this is the issue here, but when creating a new database in firebase, the default permissions are set to "allow everything for 1 week after creation". This looks something like this:
> 
> ```
> rules_version = '2';
> service cloud.firestore {
>   match /databases/{database}/documents {
>     match /{document=**} {
>       allow read, write: if request.time < timestamp.date(2020, 7, 1);
>     }
>   }
> }
> ```
> 
> You can extend the permitted time by modifying `timestamp.date(2020, 7, 1)`, or create propper permission rules.

Thanks a lot

**Comment by crevulus** [comment]
@imadelcass be careful using those rules. They're very insecure and it's not recommended.

**Comment by samuelvalleinclan** [comment]
cf

**Comment by Thokya** [comment]
> When you look at your Firestore rules, they should match:
> 
> ```
> service cloud.firestore {
>   match /databases/{database}/documents {
>     match /{document=**} {
>       allow read;
>       allow write: if false;
>     }
>   }
> }
> ```
> 
> If you do not allow read, you may see this issue in the client.

this worked for me thanks for the suggestion

**Comment by vishnusiju** [comment]
In firebase -> Cloud Firestore ->Rules,
<hr>

rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.auth != null;
    }
  }
}

<hr>
Change To,
<hr>

rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write;
    }
  }
}

**Comment by Adyaapatil09** [comment]
`rules_version = '2';
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if
          request.time < timestamp.date(2022, 12, 1);
    }
  }
}`
 you might use these rules to just remember to change the dates when they expire!

**Comment by 2003-r** [comment]
I have the same error with firebase store and I don't know how to fix it 
whenever I am trying to upload files I need to upload the file in firebase storage and fields to firebase store , it store file to the storage but fail to store in firebase store
can anyone help me ?

**Comment by elvandiano** [comment]
> In firebase -> Cloud Firestore ->Rules,
> 
> rules_version = '2'; service cloud.firestore { match /databases/{database}/documents { match /{document=**} { allow read, write: if request.auth != null; } } }
> 
> Change To,
> rules_version = '2'; service cloud.firestore { match /databases/{database}/documents { match /{document=**} { allow read, write; } } }

its work to me.. thanks guys
